digitalocean/resource_digitalocean_app.go

```diff
@@ -0,0 +1,199 @@
+			},
+		},
+	}
+}
+
+func resourceDigitalOceanAppCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*CombinedConfig).godoClient()
+	appCreateRequest := &godo.AppCreateRequest{}
+	appCreateRequest.Spec = expandAppSpec(d.Get("spec").([]interface{}))
+
+	log.Printf("[DEBUG] App create request: %#v", appCreateRequest)
+	app, _, err := client.Apps.Create(context.Background(), appCreateRequest)
+	if err != nil {
+		return fmt.Errorf("Error creating App: %s", err)
+	}
+
+	d.SetId(app.ID)
+	log.Printf("[DEBUG] Waiting for app (%s) deployment to become active", app.ID)
+
+	err = waitForAppDeployment(client, app.ID)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[INFO] App created, ID: %s", d.Id())
+
+	return resourceDigitalOceanAppRead(d, meta)
+}
+
+func resourceDigitalOceanAppRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*CombinedConfig).godoClient()
+
+	app, resp, err := client.Apps.Get(context.Background(), d.Id())
+	if err != nil {
+		if resp != nil && resp.StatusCode == 404 {
+			log.Printf("[DEBUG] App (%s) was not found - removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error reading App: %s", err)
+	}
+
+	d.SetId(app.ID)
+	d.Set("default_ingress", app.DefaultIngress)
+	d.Set("live_url", app.LiveURL)
+	d.Set("active_deployment_id", app.ActiveDeployment.ID)
+	d.Set("updated_at", app.UpdatedAt.UTC().String())
+	d.Set("created_at", app.CreatedAt.UTC().String())
+
+	if err := d.Set("spec", flattenAppSpec(app.Spec)); err != nil {
+		return fmt.Errorf("[DEBUG] Error setting app spec: %#v", err)
+	}
+
+	return err
+}
+
+func resourceDigitalOceanAppUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*CombinedConfig).godoClient()
+
+	if d.HasChange("spec") {
+		appUpdateRequest := &godo.AppUpdateRequest{}
+		appUpdateRequest.Spec = expandAppSpec(d.Get("spec").([]interface{}))
+
+		app, _, err := client.Apps.Update(context.Background(), d.Id(), appUpdateRequest)
+		if err != nil {
+			return fmt.Errorf("Error updating app (%s): %s", d.Id(), err)
+		}
+
+		log.Printf("[DEBUG] Waiting for app (%s) deployment to become active", app.ID)
+		err = waitForAppDeployment(client, app.ID)
+		if err != nil {
+			return err
+		}
+
+		log.Printf("[INFO] Updated app (%s)", app.ID)
+	}
+
+	return resourceDigitalOceanAppRead(d, meta)
+}
+
+func resourceDigitalOceanAppDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*CombinedConfig).godoClient()
+
+	log.Printf("[INFO] Deleting App: %s", d.Id())
+	_, err := client.Apps.Delete(context.Background(), d.Id())
+	if err != nil {
+		return fmt.Errorf("Error deletingApp: %s", err)
+	}
+
+	d.SetId("")
+	return nil
+}
+
+func waitForAppDeployment(client *godo.Client, id string) error {
+	tickerInterval := 10 //10s
+	timeout := 1800      //1800s, 30min
+	n := 0
+
+	// No deployment seems to be available right way, sleep 10 seconds
+	time.Sleep(10 * time.Second)
```

Comment:
@kamaln7 - Any hints on a better way to handle this? How long should it be before an in-progress deployment becomes available after a POST?

Reply:
When hitting CreateApp or UpdateApp, the new deployment will be created in the same request so you can skip this altogether and hit ListDeployments right away (unless the UpdateApp call doesn't trigger a new deployment).

The deployment moving into the in-progress state happens async so if you specifically need that I don't think there's a better way than what you already have in here. In that case you should move the `Apps.Get()` call inside the loop and check for `app.InProgressDeployment != nil`. Maybe lower the interval to `5s` or so and bump it back up to `10s` once the deployment is in progress

Reply:
Thanks! Is there any way to know that an updated _did not_ trigger a deployment beyond the absence of a new deployment?

Reply:
Sorry, I took a closer look and these cases all seem to be internal. It's safe to assume that hitting the UpdateApp endpoint will always trigger a new deployment